### PR TITLE
FIX: Proper brain-size correction in percentile

### LINKED ISSUE #822: space_net._crop_mask returns incomprehensible error when mask is empty
I have a user getting an error that he does not understand at all when the mask extraction has failed:

<pre>
C:\Users\LYF\AppData\Local\Enthought\Canopy\User\lib\site-packages\nilearn-0.1
.5.dev0-py2.7.egg\nilearn\decoding\space_net.pyc in path_scores(solver, X, y,
mask, alphas, l1_ratios, train, test, solver_params, is_classif, n_alphas,
eps, key, debias, Xmean, screening_percentile, verbose)
    356
    357     # crop the mask to have a tighter bounding box
--> 358     mask = _crop_mask(mask)
    359
    360     # get train and test data

C:\Users\LYF\AppData\Local\Enthought\Canopy\User\lib\site-packages\nilearn-0.1
.5.dev0-py2.7.egg\nilearn\decoding\space_net.pyc in _crop_mask(mask)
     61     the same support (active voxels)"""
     62     idx = np.where(mask)
---> 63     i_min = max(idx[0].min() - 1, 0)
     64     i_max = idx[0].max()
     65     j_min = max(idx[1].min() - 1, 0)

C:\Users\LYF\AppData\Local\Enthought\Canopy\User\lib\site-packages\numpy\core\
_methods.pyc in _amin(a, axis, out, keepdims)
     27
     28 def _amin(a, axis=None, out=None, keepdims=False):
---> 29     return umr_minimum(a, axis, None, out, keepdims)
     30
     31 def _sum(a, axis=None, dtype=None, out=None, keepdims=False):

ValueError: zero-size array to reduction operation minimum which has no
identity
</pre>

We should detect that "idx" is empty in _crop_mask and raise a more comprehensible error: something like "Empty mask: if you have given a mask, it is empty, and if you have not given a mask, the mask-extraction routines have fail. Please provide an appropriate mask".

@dohmatob : do you want to take care of that?

### PR DESCRIPTION
**Pull request by GaelVaroquaux**
The ratio of brain size to mask volume was applied wrong in the
screening percentile.

In addition, add a bit more warnings.

Fixes #822

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
This is a fix for #822: it turns out that the logic was wrong.

@dohmatob @bthirion @AlexandreAbraham please review: a user is waiting for this fix to be merged.

**Comment by GaelVaroquaux** [comment]
As a side note, I have checked that in the plot_haxby_spacenet example, when using as a mask the mask of the ventral temporal stream, nothing bad happens (aka the example gives good prediction and good maps).

**Comment by dohmatob** [comment]
`from sklearn.utils.testing import assert_warns` breaks in `sklearn 0.13`.

**Comment by GaelVaroquaux** [comment]
> from sklearn.utils.testing import assert_warns breaks in sklearn 0.13.

Indeed. I am on it. Aside from this, any issue with the PR.

File: nilearn/decoding/tests/test_space_net.py [hunk 1]
```diff
@@ -12,11 +12,12 @@
 from sklearn.linear_model import Lasso
 from sklearn.utils import check_random_state
 from sklearn.linear_model import LogisticRegression
+from sklearn.utils.testing import assert_warns
```
Comment (dohmatob) [inline nilearn/decoding/tests/test_space_net.py]: This import breaks in `sklearn 0.13`.

File: nilearn/decoding/space_net.py [hunk 2]
```diff
@@ -533,7 +533,8 @@ class BaseSpaceNet(LinearModel, RegressorMixin, CacheMixin):
         100 means 'keep all features'. This percentile is is expressed
         w.r.t the volume of a standard (MNI152) brain, and so is corrected
         at runtime to correspond to the volume of the user-supplied mask
-        (which is typically smaller).
+        (which is typically smaller). If '100' is given, all the features
+        are used, regardless of voxel size.
```
Comment (dohmatob) [inline nilearn/decoding/space_net.py]: change to " ... regardless of the number of voxels." or something like that.

File: nilearn/decoding/space_net.py [hunk 3]
```diff
@@ -806,9 +807,22 @@ def fit(self, X, y):
...
-                "Brain mask is bigger than volume of standard brain!")
-        self.screening_percentile_ = self.screening_percentile * (
-            mask_volume / MNI152_BRAIN_VOLUME)
+                "Brain mask is bigger than the volume of a standard "
+                "humain brain. Strange things might happen.",
+                stacklevel=2)
+        elif mask_volume < .005 * MNI152_BRAIN_VOLUME:
+            warnings.warn(
+                "Brain mask is smaller than .5% of the volume "
+                "of a standard human brain. Strange things might happen.",
```
Comment (dohmatob) [inline nilearn/decoding/space_net.py]: I would rephrase to something like ".. of a standard human brain. You may want to increase the screening percentile."

File: nilearn/decoding/tests/test_space_net.py (hunk 1 above)
Comment (GaelVaroquaux) [inline nilearn/decoding/tests/test_space_net.py]: > This import breaks in sklearn 0.13.

Fix pushed. Let's see if travis is happy.

File: nilearn/decoding/space_net.py (hunk 3 above)
Comment (GaelVaroquaux) [inline nilearn/decoding/space_net.py]: > I would rephrase to something like ".. of a standard human brain. You
> may want to increase the screening percentile."

This warning is unrelated to screening percentile. I just want to raise
the message that people are using the package in a weird way.

Comment (dohmatob) [inline nilearn/decoding/space_net.py]: ok, u're right. Still, I'd rephrase the " strange things might happen."

Comment (GaelVaroquaux) [inline nilearn/decoding/space_net.py]: > ok, u're right. Still, I'd rephrase the " strange things might happen."

Maybe something like 'SpaceNet is probably not tuned to be used on such
data'?

**Comment by dohmatob** [comment]
OK, give me 5 minutes to run a few experiments.

File: nilearn/decoding/space_net.py [hunk 4]
```diff
@@ -806,9 +807,22 @@ def fit(self, X, y):
...
+                stacklevel=2)
+
+        screening_percentile = self.screening_percentile
+        if screening_percentile < 100:
+            screening_percentile = screening_percentile * (
+                MNI152_BRAIN_VOLUME / mask_volume)
+            screening_percentile = max(self.screening_percentile, 100)
```
Comment (dohmatob) [inline nilearn/decoding/space_net.py]: This should be `screening_percentile = min(self.screening_percentile, 100)`.

Comment (GaelVaroquaux) [inline nilearn/decoding/space_net.py]: > This should be screening_percentile = min(self.screening_percentile, 100).

Good catch!

**Comment by dohmatob** [comment]
I'd wrap the whole mechanism for correcting the screening percentile into a separate _method, in order to render it testable (without fitting a model, etc.).

**Comment by dohmatob** [comment]
LGTM. +1 for merge.

File: nilearn/decoding/space_net.py (hunk 3 above)
Comment (dohmatob) [inline nilearn/decoding/space_net.py]: It's not an easy business to find statements explaining negative borderline scenarios. I'm having a hard time finding a good proposal. I'll go with what you settle on :) We can always revise this in future.

File: nilearn/decoding/space_net.py [hunk 5]
```diff
@@ -806,9 +807,22 @@ def fit(self, X, y):
         mask_volume = _get_mask_volume(self.mask_img_)
         if mask_volume > MNI152_BRAIN_VOLUME:
             warnings.warn(
-                "Brain mask is bigger than volume of standard brain!")
-        self.screening_percentile_ = self.screening_percentile * (
-            mask_volume / MNI152_BRAIN_VOLUME)
+                "Brain mask is bigger than the volume of a standard "
+                "humain brain. Strange things might happen.",
+                stacklevel=2)
+        elif mask_volume < .005 * MNI152_BRAIN_VOLUME:
```
Comment (dohmatob) [inline nilearn/decoding/space_net.py]: Or maybe if this test passes, then check the `screening_percentile` and force it to `100` if less (with the appropriate warning) ?

**Comment by GaelVaroquaux** [comment]
I've addressed your remarks, @dohmatob . Thanks for the review.

**Comment by GaelVaroquaux** [comment]
Tests pass. Can I haz one more review?

**Comment by AlexandreAbraham** [comment]
LGTM

**Comment by GaelVaroquaux** [comment]
Merging
